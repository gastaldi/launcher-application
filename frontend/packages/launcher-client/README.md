# launcher-client [![Build Status](https://travis-ci.com/fabric8-launcher/launcher-client.svg?branch=master)](https://travis-ci.com/fabric8-launcher/launcher-client)
Typescript / Javascript client to communicate with the launcher backends
