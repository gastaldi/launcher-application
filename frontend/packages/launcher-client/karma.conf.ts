module.exports = (config) => {
  config.set({
    frameworks: ['jasmine', 'karma-typescript'],
    files: [
      { pattern: 'src/**/*.ts' }
    ],
    preprocessors: {
      '**/*.ts': ['karma-typescript']
    },
    karmaTypescriptConfig: {
      tsconfig: './tsconfig.json'
    },
    reporters: ['mocha', 'karma-typescript'],
    browsers: ['ChromeHeadless'],
    singleRun: true
  });
};
