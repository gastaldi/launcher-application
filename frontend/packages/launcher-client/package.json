{
  "name": "launcher-client",
  "version": "1.0.0-alpha.0",
  "homepage": "https://github.com/fabric8-launcher/launcher-client",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/fabric8-launcher/launcher-client.git"
  },
  "bugs": {
    "url": "https://github.com/fabric8-launcher/launcher-client/issues"
  },
  "main": "build/index.js",
  "module": "build/esm/index.js",
  "types": "build/types/index.d.ts",
  "engines": {
    "node": ">= 8.9.0",
    "npm": ">= 5.5.0"
  },
  "scripts": {
    "build": "rm -rf ./build && rollup -c",
    "test": "jest",
    "semantic-release": "semantic-release"
  },
  "dependencies": {
    "axios": "0.19.0",
    "injection-js": "2.2.1",
    "lscache": "1.3.0",
    "reflect-metadata": "0.1.13"
  },
  "devDependencies": {
    "@types/axios": "0.14.0",
    "@types/lscache": "1.0.29",
    "@types/node": "11.13.4",
    "axios-mock-adapter": "1.16.0",
    "rollup-plugin-json": "4.0.0",
    "rollup-plugin-strip-code": "0.2.6"
  },
  "eslintConfig": {
    "extends": "react-app"
  }
}
