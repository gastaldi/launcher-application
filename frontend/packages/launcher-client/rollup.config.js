import peerDepsExternal from 'rollup-plugin-peer-deps-external';
import filesize from 'rollup-plugin-filesize';
import tslint from 'rollup-plugin-tslint';
import typescript from 'rollup-plugin-typescript2';
import json from 'rollup-plugin-json';

import pkg from './package.json';

const config = {
    input: 'src/index.ts',
    output: [
        {
            file: pkg.main,
            format: 'cjs',
        },
        {
            file: pkg.module,
            format: 'es',
        },
    ],
    external: [
        ...Object.keys(pkg.dependencies || {}),
        ...Object.keys(pkg.peerDependencies || {}),
    ],
    plugins: [
        peerDepsExternal(),
        tslint({
            include: './src/**/*.tsx?'
        }),
        typescript({
            clean: true,
            typescript: require('typescript'),
            verbosity: 0,
            useTsconfigDeclarationDir: true
        }),
        json(),
        // FIXME find another solution to remove the mock (that make it possible to choose from the consumer side)
        /**
        isProduction && stripCode({
            start_comment: 'test-code',
            end_comment: 'end-test-code'
        }),**/
        filesize(),
    ],
};

export default config;
