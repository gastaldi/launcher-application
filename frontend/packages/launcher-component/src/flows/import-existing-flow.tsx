import * as React from 'react';
import { generate } from 'project-name-generator';
import { useSessionStorageWithObject } from 'react-use-sessionstorage';

import { buildDownloadImportAppPayload, buildLaunchImportAppPayload } from './launcher-client-adapters';
import { SrcRepositoryHub } from '../hubs/src-repository-hub';
import { LaunchFlow, useAutoSetCluster, NAME_REGEX } from './launch-flow';
import { DeploymentHub } from '../hubs/deployment-hub';
import { ImportApp } from './types';
import { ProjectNameInput } from '../core/project-name-input/project-name-input';

const DEFAULT_IMPORT_APP = {
  srcRepository: {},
  deployment: {},
};

function getFlowStatus(app: ImportApp) {
  if (!NAME_REGEX.test(app.name)) {
    return {
      hint: 'You should enter a valid name for your application',
      isReadyForDownload: false,
      isReadyForLaunch: false
    };
  }
  if (!SrcRepositoryHub.checkCompletion(app.srcRepository)) {
    return {
      hint: 'You should configure the source repository.',
      isReadyForDownload: false,
      isReadyForLaunch: false,
    };
  }
  if (!DeploymentHub.checkCompletion(app.deployment)) {
    return {
      hint: 'If you wish to Launch your application, you should configure OpenShift Deployment.',
      isReadyForDownload: true,
      isReadyForLaunch: false,
    };
  }
  return {
    hint: 'Your application is ready to launch!',
    isReadyForDownload: true,
    isReadyForLaunch: true,
  };
}

export function ImportExistingFlow(props: { appName?: string; onCancel?: () => void }) {
  const defaultAppName = props.appName || generate().dashed;
  const defaultImportApp = {...DEFAULT_IMPORT_APP, name: defaultAppName};
  const [app, setApp, clear] = useSessionStorageWithObject<ImportApp>('import-existing-app', defaultImportApp);
  const onCancel = () => {
    clear();
    props.onCancel!();
  };
  const showDeploymentForm = useAutoSetCluster(setApp);

  const flowStatus = getFlowStatus(app);

  const items = [
    {
      id: SrcRepositoryHub.id,
      title: SrcRepositoryHub.title,
      overview: {
        component: ({edit}) => (
          <SrcRepositoryHub.Overview value={app.srcRepository} onClick={edit}/>
        ),
        width: 'half',
      },
      form: {
        component: ({close}) => (
          <SrcRepositoryHub.Form
            initialValue={app.srcRepository}
            onSave={(srcRepository) => {
              setApp(prev => ({...prev, srcRepository}));
              close();
            }}
            onCancel={close}
          />
        ),
      }
    },
    {
      id: DeploymentHub.id,
      title: DeploymentHub.title,
      overview: {
        component: ({edit}) => (
          <DeploymentHub.Overview value={app.deployment} onClick={edit}/>
        ),
        width: 'half',
      },
      form: showDeploymentForm && {
        component: ({close}) => (
          <DeploymentHub.Form
            initialValue={app.deployment}
            onSave={(deployment) => {
              setApp(prev => ({...prev, deployment}));
              close();
            }}
            onCancel={close}
          />
        ),
      }
    }
  ];

  return (
    <LaunchFlow
      title={(
        <ProjectNameInput
          prefix="Import an Existing Application as:"
          value={app.name}
          onChange={value => setApp(prev => ({...prev, name: value}))}
        />
      )}

      items={items}
      {...flowStatus}
      buildLaunchAppPayload={() => {
        clear();
        return buildLaunchImportAppPayload(app);
      }}
      buildDownloadAppPayload={() => {
        clear();
        return buildDownloadImportAppPayload(app);
      }}
      onCancel={onCancel}
    />
  );

}
