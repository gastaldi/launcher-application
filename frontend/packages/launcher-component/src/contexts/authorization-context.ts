import React, { useContext } from 'react';
import { checkNotNull, Authorizations } from 'launcher-client';

export interface AuthorizationsManager {
    getAuthorizations: (provider: string) => Promise<Authorizations | undefined>;
    generateAuthorizationLink(provider?: string, redirect?: string): string;
}

export const AuthorizationManagerContext = React.createContext<AuthorizationsManager | undefined>(undefined);

export function useAuthorizationManager(): AuthorizationsManager {
  const auth = useContext(AuthorizationManagerContext);
  return checkNotNull(auth, 'Authorization must be defined in context');
}
