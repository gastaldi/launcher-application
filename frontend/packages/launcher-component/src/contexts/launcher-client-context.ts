import * as React from 'react';
import { useContext } from 'react';
import { checkNotNull, LauncherClient } from 'launcher-client';

export const LauncherClientContext = React.createContext<LauncherClient | undefined>(undefined);

export function useLauncherClient(): LauncherClient {
  const client = useContext(LauncherClientContext);
  return checkNotNull(client, 'launcher client must be defined in context');
}
