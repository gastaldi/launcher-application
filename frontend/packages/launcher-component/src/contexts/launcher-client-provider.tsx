import React, { useState } from 'react';
import { LauncherClientContext } from './launcher-client-context';
import { cachedLauncherClient, checkNotNull, LauncherClient, mockLauncherClient } from 'launcher-client';
import { AuthorizationsManager, AuthorizationManagerContext } from './authorization-context';

interface LauncherDepsProvider {
  children: React.ReactNode;
  client?: LauncherClient;
  authorizationsManager?: AuthorizationsManager;
  creatorUrl?: string;
  launcherUrl?: string;
}

function buildLauncherClient(props: LauncherDepsProvider) {
  if (props.client) {
    return props.client;
  }
  let client: LauncherClient;
  if (!!props.creatorUrl || !!props.launcherUrl) {
    checkNotNull(props.launcherUrl, 'launcherUrl');
    checkNotNull(props.creatorUrl, 'creatorUrl');
    client = cachedLauncherClient({ creatorUrl: props.creatorUrl!, launcherURL: props.launcherUrl! });
  } else {
    client = mockLauncherClient();
  }
  return client;
}

function buildAuthorizationManager(props: LauncherDepsProvider): AuthorizationsManager {
  if (props.authorizationsManager) {
    return props.authorizationsManager;
  }
  return {
    getAuthorizations: () => Promise.resolve({}),
    generateAuthorizationLink: (provider?: string) => `http://mock-authorization/${provider}`
  };
}

export function LauncherDepsProvider(props: LauncherDepsProvider) {
  const [client] = useState<LauncherClient>(buildLauncherClient(props));
  const [authorizationManager] = useState<AuthorizationsManager>(buildAuthorizationManager(props));
  client.authorizationsProvider = authorizationManager.getAuthorizations;

  return (
    <LauncherClientContext.Provider value={client}>
      <AuthorizationManagerContext.Provider value={authorizationManager}>
        {props.children}
      </AuthorizationManagerContext.Provider>
    </LauncherClientContext.Provider>
  );
}
