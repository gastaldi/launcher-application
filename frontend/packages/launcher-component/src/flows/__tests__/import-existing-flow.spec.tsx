import * as React from 'react';
import 'jest-dom/extend-expect';
import { cleanup, fireEvent, render } from 'react-testing-library';
import { ImportExistingFlow } from '../import-existing-flow';
import { mockLauncherClient } from 'launcher-client';
import { launchCheckPayloadAndProgress } from './flow-helpers';
import { LauncherDepsProvider } from '../../contexts/launcher-client-provider';
import { flushPromises } from '../../core/__tests__/test-helpers';

afterEach(() => {
  console.log('cleanup()');
  cleanup();
});

jest.useFakeTimers();

async function configureSrc(comp, url) {
  fireEvent.click(comp.getByLabelText('Open src-repository editor'));
  expect(comp.getByLabelText('Edit src-repository')).toBeDefined();

  fireEvent.change(comp.getByLabelText('Git repository url'), { target: { value: url } });
  fireEvent.click(comp.getByText('Done'));

  // Resolve fetch url
  await flushPromises();

  fireEvent.click(comp.getByText('Advanced settings'));
  expect(comp.getByLabelText('select-buildImage')).toBeDefined();

  fireEvent.click(comp.getByLabelText('Choose Java Code Builder'));

  fireEvent.change(comp.getByPlaceholderText('Type the environment variable name'), {target: {value: 'JAVA_DEBUG'}});
  fireEvent.change(comp.getByPlaceholderText('Type the environment variable value'), {target: {value: 'true'}});
  fireEvent.click(comp.getByLabelText('Save src-repository'));

  // Resolve overview promises
  await flushPromises();
}

describe('<ImportExistingFlow />', () => {
  it('renders and initializes the ImportExistingFlow correctly', async () => {
    const comp = render(<LauncherDepsProvider><ImportExistingFlow appName="my-test-app" /></LauncherDepsProvider>);
    expect(comp.getByLabelText('openshift-deployment is not configured')).toBeDefined();

    // Resolve data from auto loader
    await flushPromises();

    // Resolve overview promises
    await flushPromises();

    expect(comp.getByLabelText('openshift-deployment is configured')).toBeDefined();

    expect(comp.getByLabelText('Launch Application')).toHaveAttribute('disabled');
    expect(comp.getByLabelText('Download Application')).toHaveAttribute('disabled');
  });

  it('Configure source repository to import and check full launch until next steps popup', async () => {
    const mockClient = mockLauncherClient();
    const comp = render(<LauncherDepsProvider client={mockClient}><ImportExistingFlow appName="my-test-app" /></LauncherDepsProvider>);

    // Resolve data from auto loader
    await flushPromises();

    // Resolve overview promises
    await flushPromises();

    await configureSrc(comp, 'https://github.com/nodeshift-starters/nodejs-rest-http');
    expect(comp.getByText('Import is configured')).toBeDefined();

    expect(comp.getByLabelText('Launch Application')).not.toHaveAttribute('disabled');
    expect(comp.getByLabelText('Download Application')).not.toHaveAttribute('disabled');

    await launchCheckPayloadAndProgress(comp, mockClient);
  });
});
