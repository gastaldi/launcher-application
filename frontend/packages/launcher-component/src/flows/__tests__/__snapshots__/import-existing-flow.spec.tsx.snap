// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImportExistingFlow /> Configure source repository to import and check full launch until next steps popup: Console link 1`] = `"https://console.starter-us-east-2.openshift.com/console/projects"`;

exports[`<ImportExistingFlow /> Configure source repository to import and check full launch until next steps popup: payload 1`] = `
Object {
  "clusterId": "starter-ca-central-1",
  "project": Object {
    "application": "my-test-app",
    "parts": Array [
      Object {
        "capabilities": Array [
          Object {
            "module": "import",
            "props": Object {
              "builderImage": "registry.access.redhat.com/redhat-openjdk-18/openjdk18-openshift",
              "env": Object {
                "JAVA_DEBUG": "true",
              },
              "gitImportUrl": "https://github.com/nodeshift-starters/nodejs-rest-http",
            },
          },
        ],
        "category": "import",
        "shared": Object {},
      },
    ],
  },
  "projectName": "my-test-app",
}
`;
